### PR TITLE
Java bindings for Parquet struct support 

### PR DESCRIPTION
**Pull request by razajafri**
This PR adds support for writing Structs to Parquet.

### CODE REVIEW

**Comment by razajafri** [comment]
@revans2 I have removed the params from the method docs and I have made some more changes to help in writing the plugin code. PTAL at the last commit called "updated API", everything before that has been reviewed already

**Comment by razajafri** [comment]
rerun tests

File: cpp/include/cudf/io/parquet.hpp
Comment (harrism) [inline cpp/include/cudf/io/parquet.hpp:394]: No other data members are in this class are public. Instead, they have getters and setters. Why did you make `children` public?

**Review by harrism — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/io/parquet.hpp
Comment (razajafri) [reply cpp/include/cudf/io/parquet.hpp:394]: I am not a cpp expert so please let me know if there is a better way to do this. 

The way @devavret has designed the API works perfectly when we have a table and we initialize the table_input_metadata with it, if we don't then the children vector is never initialized this is why I made it public so I can initialize it.

Comment (harrism) [reply cpp/include/cudf/io/parquet.hpp:394]: You are violating encapsulation. @devavret or @vuule who know these classes better should review.

Comment (razajafri) [reply cpp/include/cudf/io/parquet.hpp:394]: Thanks, I have spoken to @devavret about this but I would love for him or @vuule to review.

Comment (devavret) [reply cpp/include/cudf/io/parquet.hpp:394]: This was allowed for the time being to unblock the API development. It's how the old metadata class functioned. I'm thinking now whether adding a `add_column()` method would be useful. It would still need to construct its entire hierarchy at once.

Comment (harrism) [reply cpp/include/cudf/io/parquet.hpp:394]: At least add a comment. This really should be done the right way, it wouldn't be hard to add a `get_children` accessor...

Comment (razajafri) [reply cpp/include/cudf/io/parquet.hpp]: @harrism I have added a getter for the private member.

**Review by devavret — COMMENTED** [review]
Approving to unblock for now. Will probably revisit.

File: cpp/include/cudf/io/parquet.hpp
Comment (vuule) [inline cpp/include/cudf/io/parquet.hpp:399]: Accessor returning a mutable reference is not much better than direct access. Given the usage, I would recommend replacing this with something like `add_child(column_in_metadata const&)`. The call site would not change too much, and we avoid temporarily having default-initialized columns in the vector.

**Review by vuule — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/io/parquet.hpp
Comment (razajafri) [reply cpp/include/cudf/io/parquet.hpp]: That's exactly what I was thinking but I did it anyway. I have addressed your concern PTAL

**Review by vuule — APPROVED** [review]
Looks good, thank you!
As a side note, we'll probably improve the API later on (e.g. enable adding a child without declaring it as a variable).

**Comment by razajafri** [comment]
> Looks good, thank you!
> As a side note, we'll probably improve the API later on (e.g. enable adding a child without declaring it as a variable).

That's fine, just give us a heads up if its a breaking change so we aren't caught off-guard

**Comment by vuule** [comment]
> > Looks good, thank you!
> > As a side note, we'll probably improve the API later on (e.g. enable adding a child without declaring it as a variable).
> 
> That's fine, just give us a heads up if its a breaking change so we aren't caught off-guard

I'm hoping this would just be an addition. Will make sure to notify otherwise.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:250]: This ends up creating a new temporary array of the result for each child and at every level.  Gathering all the child column name results and then concatenate them together at the end, when the final array size for this level can be computed directly, will generate less garbage and reduce the number of array copies.  This comment also applies to the other places where this algorithm was replicated.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:40]:
```suggestion
#include <cudf/types.hpp>
#include <rmm/cuda_stream_view.hpp>
```

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:1115]: Nit: These internal functions should be in an anonymous namespace and moved outside of the `extern "C"` block.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:1122]: I found it confusing that we're _not_ reading the names, etc. from the specified read_index but actually read_index + 1 because of this.  It might be clearer if `num_children` is passed into this method since every caller is already loading it to decide whether to call this, and then the algorithm can start reading names, etc. from `read_index` as expected.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:1157]:
```suggestion
  // this should never be 0, because a table must have columns
  int top_level_children = children[0];
```

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:1160]: I found this comment a bit confusing, can this elaborate a bit more?  Maybe if the number of children are passed in as an argument we wouldn't have dummy entries and make this easier to follow?

File: java/src/test/java/ai/rapids/cudf/TableTest.java
Comment (jlowe) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:4791]: nit: typo: `ninth` instead of `nineth` here and elsewhere

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:142]: Should this be `withNullableColumns` so it's clearer that multiple arguments are supported? The name implies singular.  This comment applies to other places where varargs are supported.

**Comment by razajafri** [comment]
@jlowe I have addressed your concerns PTAL

**Review by harrism — APPROVED** [review]
Thanks for addressing my concern.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:257]: This should short-circuit if there are no child columns (which is going to be a common case).

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:259]: Rather than needing to remember to tack on `ret.length` later, we can initialize the total length to `ret.length` here.  Otherwise it's a bit weird that the variable name implies it's the total length when it's actually not.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:291]: Same short-circuit comment here.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:293]: Same "make it a true total length" comment here.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:315]: Same short-circuit comment here.

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:317]: Same "make it a true total length" comment here.

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:668]: My comment from https://github.com/rapidsai/cudf/pull/7998#discussion_r622431512 didn't get addressed.

Comment (razajafri) [reply java/src/main/native/src/TableJni.cpp]: I got mixed up with the comment for passing the top_level_children as argument from Java. I have addressed it now.

**Review by revans2 — COMMENTED** [review]
Just some nits

**Review by revans2 — APPROVED** [review]

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:680]: The point of passing the number of children as an argument is to simplify this logic as well, e.g.:
```suggestion
    int child_num_children = children[read_index++];
    if (child_num_children > 0) {
      read_index = set_column_metadata(column_metadata.child(write_index), col_names, nullability,
                                       isInt96, precisions, children, child_num_children, read_index);
    }
```

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:716]:
```suggestion
    int child_num_children = children[read_index++];
    if (child_num_children > 0) {
      read_index = set_column_metadata(metadata.column_metadata[write_index], cpp_names,
                                       col_nullability, isInt96, precisions, children, child_num_children, read_index);
    }
```

File: java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java:270]:
```suggestion
    boolean[] result = new boolean[totalChildrenFlatLength];
```

**Review by jlowe — CHANGES_REQUESTED** [review]

File: java/src/main/native/src/TableJni.cpp
Comment (jlowe) [inline java/src/main/native/src/TableJni.cpp:712]:
```suggestion
    int childs_children = children[read_index];
```

**Review by jlowe — APPROVED** [review]
Minor formatting nit but otherwise lgtm.

**Comment by razajafri** [comment]
Waiting on https://github.com/rapidsai/cudf/pull/8114 to be merged

**Comment by razajafri** [comment]
rerun tests

**Comment by razajafri** [comment]
@gpucibot merge

**Pull request merged**
